### ISSUE TITLE
Implement MT-32 emulation

### ISSUE THREAD

**Issue by dreamer** [dosbox-staging/dosbox-staging#257]
MT-32 emulation is implemented using GPLv2 by https://github.com/munt/munt - Munt also provides an ability to compile as a library (mt32emu). This feature is only about integrating mt32 emulator, *NOT* about bundling necessary ROM files, (that are copyrighted by Roland and are non-free software).

Patch for implementing it is distributed in Munt ([link](https://github.com/munt/munt/tree/master/DOSBox-mt32-patch)), but also e.g. in DOSBox ECE: bdcf933ae9989515d6381d1907077db273c220dd. However, these patches assume that users will compile the mt32emu library themselves, as libmt32emu is not distributed via any software repositories (no vcpkg, no brew, almost no Linux repo presence), which is not a long-term solution.

dosbox-staging will go a similar route as ScummVM already did (bundle Munt code in this repo), but details are not decided yet (as source code dump? as Git submodule? as Git subtree?). The prerequisite of this is the modernization of buildsystem (dosbox-staging uses autotools right now, Munt uses cmake).

---
**edit**: *Update 2020-12-18: patch is now merged to master, users can enable the feature during compilation with --enable-mt32; see the newest comments for current progress on this feature*

**Comment by bluddy** [comment]
Will you also add fluidsynth support? Fluidsynth and mt32 support are the 2 things keeping me on dosbox-ece.

**Comment by dreamer** [comment]
@bluddy #262 :)

**Comment by i30817** [comment]
libmt32 can be built as a static library for compile time only linking.

It didn't save me from creating a recipe to build it for my ppa, but just having the dosbox mt32 patch available by sergm all the time every time always uptodate is incredible.

https://code.launchpad.net/~i30817/+recipe/mt32emu-daily-for-dosbox

**Comment by dreamer** [comment]
@i30817 GUI is not important. libmt32 bundled with dosbox code means it will be built as a static library - like every other code module inside dosbox-staging.

The tricky part is: doing this while keeping ease of build for all new developers, distributors, etc, maintaining coherent buildsystem runnable in various environments, running code with MT32 feature through our CI and static analysis, etc. This would be a lot easier if libmt32 was packaged by various repositories - but it's not, there's not enough users of that project as a library.

Any user can just build munt as a library right now, apply patch to dosbox-staging, and get this feature this way - that's not the hard part. Someone needs to put in work make it maintainable for future, document it, test if it does not introduce regressions, *review the code*, improve the code if there are leaks, bugs, or other issues, etc.

If you have experience with it, go ahead and start chipping on the problem :).

**Comment by kcgen** [comment]
> Someone needs to put in work make it maintainable for future, document it, test if it does not introduce regressions, review the code, improve the code if there are leaks, bugs, or other issues, 

(One more) put a thread around the entire library and create and elegant shim to allow  safe and low-latency communication between DOSBox's midi API and the munt input API.

**Comment by i30817** [comment]
There originally were two patches on sergm's repo, one threaded but the functionality was built in into both patches.


https://github.com/munt/munt/tree/master/DOSBox-mt32-patch


just read them, the feature already exists and you turn it on with 'mt32.thread=true' on the dosbox config.

In fact the only thing i do to the patch is to apply a patch to the result of the patch after it's applied to change the default dir of a config file mt32 rom from nothing or dosbox conf directory (whatever it was i can't remember), to a system shared dir, because i ask for the rom during install to make apt manage the rom (it'll only get deleted i delete both exult and dosbox, the two packages that use the rom default dir or purge that installed package).

**Comment by dreamer** [comment]
> There originally were two patches on sergm's repo

I already imported various versions of patch, long time ago;  check branches:
```
  munt/mt32-0.74.3-r4249
  munt/mt32-r402
  po/ece-r4305-1
```
If you know any other version of this patch, that might of interest, please attach it with information about what SVN revision it is based on.

**Comment by bluddy** [comment]
I'd rather see this discussion take place after integration of the patch than before it. Why? Because there's a huge interest in making dosbox-staging the go-to choice for dosbox emulation. Currently, users go to dosbox-ece for their full-featured dosbox, despite dosbox-ece's continued messing with svn and patches. Get people to switch to dosbox-ece first. Then figure out how to improve the architecture of this patch.

**Comment by bluddy** [comment]
However, I agree that only the threaded version of the patch should be integrated.

**Comment by dreamer** [comment]
> I'd rather see this discussion take place after integration of the patch than before it. Why? (…) Then figure out how to improve the architecture of this patch.

That's essentially the same way we handled CD-DA, SDL2, and other patches :) Review, cleanup, assure no obvious regressions, perhaps split into smaller patches, merge to master, then start improvements.

I just don't want to start working on this before release branch for 0.75.0 is created (otherwise we'll risk the same situation the upstream is in, where next stable release is being postponed year after year). 

BTW, I am open to giving out maintainer-level access to this repo for people, who want to contribute - it makes it a bit easier to land change than via PR from external contributor - just saying ;). (Here's our [WIP Contributing guide](https://github.com/dosbox-staging/dosbox-staging/blob/po/contributing-2/CONTRIBUTING.md))

**Comment by kcgen** [comment]
Exciting update @davidhs from Vogons.org posted: [Munt using D-110 ROMs ](https://www.vogons.org/viewtopic.php?f=29&t=76084).

For those who are IP-banned at Vogons, content has been reproduced here:

> Hello all, I'm David. I seem to have wandered around the forums but I think now is the best time to create a new account in here 😁
> I tried to modify Munt to use the Roland D-110 ROMs dumped on the net and surprisingly it works quite well!
> Although I find a few differences from the real one that I have, for example the AcouPiano1 tone in the real module seem to have shorter decay than the one in Munt.
> I posted a YouTube video on demonstrating it, playing it's own demo song: Macho Memory:
> https://www.youtube.com/watch?v=pyk_pJlwqiU

-- Credit **@davidhs**
-- Content [is licensed](https://www.vogons.org/licenses) [CC-BY-4.0](https://spdx.org/licenses/CC-BY-4.0.html), thank you Vogons.

**Comment by MasterO2** [comment]
> 
> 
> Exciting update from @davidhs of the Vogons.org forums: [Munt using D-110 ROMs ](https://www.vogons.org/viewtopic.php?f=29&t=76084).
> 
> For those who are IP-banned at Vogons, content has been reproduced here:
> 
> > Hello all, I'm David. I seem to have wandered around the forums but I think now is the best time to create a new account in here 😁
> > I tried to modify Munt to use the Roland D-110 ROMs dumped on the net and surprisingly it works quite well!
> > Although I find a few differences from the real one that I have, for example the AcouPiano1 tone in the real module seem to have shorter decay than the one in Munt.
> > I posted a YouTube video on demonstrating it, playing it's own demo song: Macho Memory:
> > https://www.youtube.com/watch?v=pyk_pJlwqiU
> 
> -- Credit **@davidhs**
> -- Content [is licensed](https://www.vogons.org/licenses) [CC-BY-4.0](https://spdx.org/licenses/CC-BY-4.0.html), thank you Vogons.

@kcgen This feature is just one more way that dosbox-staging is going to make Vogons look really, really petty.

**Comment by IlyaIndigo** [comment]
I have not found an issue section on munt/munt project, so I am writing here.
I did not understand how dosbox integration works, or rather, where do I need to register CM32L_CONTROL.ROM and CM32L_PCM.ROM so that dosbox can see them?
I patched dosbox, compiled it with libmt32emu support, indicated in dosbox config (~/.dosbox/dosbox-0.74-3.conf)
mpu401 = intelligent
mididevice = mt32
midiconfig = 128:0
But when starting dune2 (if you don't start mt32emu-qt before), dosbox writes
MIXER: Got different values ​​from SDL:freq 48000, blocksize 64
MT32: Control ROM file not found
MIDI: Can't open device: mt32 with config:128:0.
MT32: Control ROM file not found
ALSA: Client initialised [128:0]
MIDI: Opened device: alsa
And sound. of course not.

If you run mt32emu-qt before that, then it, as without this integration, creates a 128:0 or 129:0 device if fluidsynth is running with GM and there is sound.
I thought this integration would save me from having to run mt32emu-qt before starting dosbox, but I haven't figured out how to set it up yet. I didn't find the instruction.

**Comment by dreamer** [comment]
For people patching our code by themselves to get this feature: here's a new version of mt-32 patch by @sergm, updated to SDL2 and adjusted for code refactorization implemented in dosbox-staging:

4c114197163de220604ffbec41d23003ef2516ab (branch `po/mt-32-1`) ([patch file for people who use sources via tarball](https://github.com/dosbox-staging/dosbox-staging.github.io/releases/download/artifacts/0001-Update-mt-32-patch-to-apply-cleanly-to-dosbox-stagin.patch)).

*After looking at the patch in more depth and with new knowledge after integrating FluidSynth - this patch is in quite good shape already (not good enough for integration in dosbox-staging right away though), it might be prudent to integrate it first before working on #735; won't happen for 0.76.0, but we might do it for 0.77.0.*

***I have not tested it myself** - I am trying to stay away from MT32 ROM files to allow for clean-room reverse engineering efforts if needed in the future.*

**Comment by IlyaIndigo** [comment]
>     * [patch file for people who use sources via tarball](https://github.com/dosbox-staging/dosbox-staging.github.io/releases/download/artifacts/0001-Update-mt-32-patch-to-apply-cleanly-to-dosbox-stagin.patch)

Unfortunately, it does not apply to 0.75.2.
Even if I modify it again it applies but doesn't work.
[dosbox-staging-0.75.2-mt32-build.log](https://github.com/dosbox-staging/dosbox-staging/files/5612649/dosbox-staging-0.75.2-mt32-build.log)
[dosbox-staging-0.75.2-mt32.patch.txt](https://github.com/dosbox-staging/dosbox-staging/files/5612653/dosbox-staging-0.75.2-mt32.patch.txt)

**Comment by IlyaIndigo** [comment]
In branch "po/mt-32-1" I noticed 2 problems.
1 Integrated fluidsynth does not work.
In the config I specify soundfont /usr/share/sounds/sf2/FluidR3_GM.sf2 but WarCraft is silent.
External fluidsynth with the same soundfont works fine.
2 The language file russian.txt from dosbox has ceased to work correctly (shows karakozyabi).
In dosbox-staging-0.75.2, it works fine.

**Comment by kcgen** [comment]
@IlyaIndigo , 0.75.2 is prior to the MIDI changes; the mt-32 branch is updated against the latest master branch.
If you're using git on Linux:

1. Hard reset to the master branch: `git checkout master -f`
2. Pull updates: `git pull`
3. Make a new MT-32 test branch off master: `git checkout -b kc/test-mt32-1`
4. Merge @dreamer's branch: `git merge origin/po/mt-32-1`
5. Build: `./scripts/build.sh -c gcc -t release`

![2020-11-29_11-03_1](https://user-images.githubusercontent.com/1557255/100550962-938a4c80-3232-11eb-947a-c311ae1134c8.png)
![2020-11-29_11-03](https://user-images.githubusercontent.com/1557255/100550963-95eca680-3232-11eb-9698-fe0b3e69fb3d.png)

**Comment by IlyaIndigo** [comment]
@kcgen I want to patch a release, not a git tarball, in which at least the language file is broken and fluidsynth still doesn't work.
I don’t want to send git tarball to the openSUSE Oss repo, as well as without the mt32 patch, which is for the original dosbox.

**Comment by kcgen** [comment]
@IlyaIndigo , The mt-32 branch (and derived patch) only applies to the 0.76-latest-master branch.

To be clear:
- the mt-32 branch (and derived patch) will not apply or work with prior releases, such as 0.75.x.
- Prior releases, such as 0.75.x, do not have the MIDI and FluidSynth work present on 0.76. So it's expected that FluidSynth will not work with 0.72.2.
 
You can copy and paste the above commands exactly as-is into your shell and produce a working executable if you like, without need for any tarballs.

**Comment by dreamer** [comment]
For 0.75.2 this revised patch makes no sense (as there are not FluidSynth changes nor MIDI refectoring in 0.75.2) - for that release use original patch from the Munt project.

**Comment by grapeli** [comment]
> If you're using git on Linux:
> 
> ### Munt
> Download, build, and install the latest Munt sources
> 
> 1. Clone: `git clone --depth 1 https://github.com/munt/munt`
> 2. Build: `cd munt/mt32emu && cmake . && make`
> 3. Install: `sudo make install`
> 

Mercy. **Beware of such advice.**

**Comment by kcgen** [comment]
@grapeli, this comment doesn't include reasoning, and therefore isn't adding value.

Can you replace your comment with the reason behind your warning?

I'm assuming you don't like the make **install** part, and would prefer not installing, and instead linking the munt libs statically; which is something that I would prefer.. but was short on time and effort to sort all that out.

If you do have better recommendations, please off steps too. I will gladly drop an improvement in its place.

**Comment by kcgen** [comment]
dreamer_ - I'd like to start the Munt merge. 
Should I start a new branch and cherry-pick your[ 2020-11-28 commit](cbdd9ff194a08ced342c883d377cf614e65d4c10) where you adjusted segm's patch to apply to DOSBox Staging?
From there, I would then apply commits transforming it to the cleaned up implementation?

**Comment by dreamer** [comment]
Argh, I had it nicely described and then my browser crashed :(

@kcgen start a new branch and cherry-pick the commit - it already has authorship, authorship date, my co-authorship set correctly, as well `Imported-from` frield.

After that usual work will need to commence: making it an **opt-in** feature (disabled by default), fixing warnings, etc. For now we only want the patch merged in and nothing more (this will be good enough e.g. for Gentoo to enable the feature if they want to).

For the first revision some significant work will need to happen - e.g. we need to remove `mt32options.h` file and introduce `MT32_AddConfigSection` function (modelled after `FLUID_AddConfigSection`). I think for the first revision we need to either remove or hide most of the knobs that are being exposed right now. I bet they are useful for debugging Munt, but for normal users - they are overkill.

Use feature define `C_MT32EMU`, please.

**Comment by MasterO2** [comment]
@kcgen @dreamer For MT-32, does that mean we will have to provide our own ROMs in order to use it, like with GUS?

**Comment by kcgen** [comment]
Thanks @dreamer; that's perfect. I plan to mirror how you structured FluidSynth.

Also agree on the options; we will find sweet-spot defaults that provide recommended/high-quality output (similar to FluidSynth's), and not expose them if possible

Same with passing through sample-rate from the mixer to get native-passthru of samples. We'll avoid resampling CPU overhead and slight loss of SnR (just like we've done for OPL and FluidSynth).

@MasterO2 , correct - for now we won't be bundling ROMs, so they will have to be provided by the user.

Great question though. Regarding where they will reside: we'll provide a default lookup location in our `conf` directory just like SF2s (so users can drop them in there once-and-for-all), but we will also permit a path specifier -- so packagers could bundle them hanging off a relative directory tree.

**Comment by dreamer** [comment]
@MasterO2 correction - you don't need to provide any ROM files for GUS. **GUS emulation works out of the box**, you just need to turn it on via `gus = true`.

If you want to use software MIDI emulator that was being bundled with GUS drivers, then yeah - you need to place it in ULTRASND directory (many users mistake this MIDI via GUS for real GUS - not, it's just yet another way to play MIDI music) :).

As for MT-32 - after we'll have mt32emu support merged in, we'll discuss with Munt devs the possibilities for solving the ROM issue.

**Comment by dreamer** [comment]
Update: #772 got merged into the master branch. Initial MT-32 support is now in, **but only as an opt-in feature** (for now).

Users no longer need to patch the sources - support can be enabled at compile time with:

    ./configure --enable-mt32

This will try to detect the library using pkg-config. Users on systems where libmt32emu is not available in repositories (which is great majority) need to compile it themselves for now.

*Tip: here's how I handle this on my Fedora system via toolbox (toolbox creates lightweight container for me to build in, so locally-built libmt32emu does not pollute /usr/local/ on my host machine:*

```
$ cd src  # I keep all projects in ~/src
$ git clone git@github.com:dosbox-staging/dosbox-staging.git
$ git clone git@github.com:munt/munt.git
$ toolbox enter --release 33    # this enters lightweight container for Fedora 33
$ sudo dnf install ccache
$ sudo dnf install <all dependencies for dosbox-staging, as in README.md>
$ sudo dnf install cmake        # for munt
$ cd munt/mt32emu
$ mkdir build $$ cd build
$ cmake ..
$ sudo make install     # this installs in toolbox container only (!) won't pollute your OS
$ cd ~/src/dosbox-staging
$ ./autogen.sh
$ ./configure --enable-mt32
$ make -j$(nproc)  # build will succeed but for testing you need to manually place libmt32emu.so somewhere in LD_LIBRARY_PATH!
```
Once the transition to new buildsystem will be complete (we are migrating to meson), we'll provide a way to build the feature without polluting /usr/local.

**Comment by kcgen** [comment]
**Linux users:** for an optimized **mt32emu** and **dosbox** binary, paste the following into your shell and then follower @dreamer's steps. 

``` shell
export CC="ccache gcc"
export CXX="ccache g++"
export ARFLAGS="Dcr"
export CFLAGS="-pipe -fpch-preprocess -DNDEBUG -O3 -fno-signed-zeros -fno-trapping-math -fassociative-math -frename-registers -ffunction-sections -fdata-sections -march=native"
export LDFLAGS="-Wl,--as-needed"
```

Then add the following your existing dosbox config file:
``` ini
[midi]
mididevice = mt32

[mt32]
romdir = 
```

Where **romdir** points to the directory holding your ROM files, named as follows:
 - CM32L_CONTROL.ROM
 - CM32L_PCM.ROM
 - MT32_CONTROL.ROM
 - MT32_PCM.ROM

**macOS and Windows users:** the recent migration from autotools to meson that @dreamer's kicked off is the precursor to elegantly unlocking MT-32 without resorting to hacky scripts.

**Comment by nemo93** [comment]
@kcgen thanks a lot for the mt32 integration. Also @dreamer for laying off the foundation. Looks really great. I do have a question after reading that there are [various mt32 ROMs](https://en.wikipedia.org/wiki/User:Lord_Nightmare) available with different release dates. 

On the ScummVM forum (I've lost the reference sorry) someone claimed that it's best to rely on ROMs with the older date/revision as some games actually do exploit some mt32 bugs to produce sound. 

Are there any specific md5/sha1 I have to look for to be sure the ROMs I have will work on Staging? Alternatively I could test this by myself as I have only 2 versions of CM32L_CONTROL.ROM and MT32_CONTROL.ROM.

**Comment by nemo93** [comment]
For the record, here the list of supported MT32 roms by Munt:

https://github.com/munt/munt/blob/master/mt32emu/src/ROMInfo.cpp#L28

**Comment by dreamer** [comment]
I just merged the new build system to the master branch - MT-32 feature is now enabled by default :)

Users who are missing mt32emu library on their systems (most of people) *don't need to compile it themselves any more*. Meson (our new buildsystem) will download the sources and recompile the library automatically - user has an option to decide if dynamic or static linking should be used.

MT-32 feature is **enabled** and working in macOS and Linux snapshot builds - anyone can test it.

MT-32 feature is **not enabled** in Windows release snapshot builds (lib is missing from Microsoft's vcpkg repository) - If someone has access to Windows machine and could try if Meson-generated VS solution works already (or send us patches if it does not work) - that would be appreciated.

**Comment by LowLevelMahn** [comment]
> MT-32 can then be selected by setting the following your existing dosbox config file:
> 
> ```ini
> [midi]
> mididevice = mt32
> 
> [mt32]
> romdir = 
> ```
> 
> Where **romdir** points to the directory holding your ROM files, named as follows:
> 
> * CM32L_CONTROL.ROM
> * CM32L_PCM.ROM
> * MT32_CONTROL.ROM
> * MT32_PCM.ROM

this way the CM32L gets loaded - the MT32 files get only loaded if there are no CM32L files
better split the MT32_* and CM32L_* in a MT32 and CM32 subfolder

**Comment by kcgen** [comment]
If you're building from source, then I suggest using the master branch, which gives you detailed model selection (will be described in the config file it generates for you, if you delete your current one). 

See thisbPR for some notes:

https://github.com/dosbox-staging/dosbox-staging/pull/965

And this PR explains how you will see your selected model on the console on startup:

https://github.com/dosbox-staging/dosbox-staging/pull/833

I also suggest getting the 'mame mt32 roms' and 'mame cm32l roms' (Google will help!), Unzip them as-is alongside your existing roms.

See this page: https://en.wikipedia.org/wiki/List_of_MT-32-compatible_computer_games#Table_of_compatible_games

Stunts will need `model = mt32_old`.

Curious how it goes!

**Comment by IlyaIndigo** [comment]
> README.md
> MT-32 emulator | Yes＊ (libmt32emu 2.4.2) | **N/A**

In DosBOX MT32 is supported via a patch provided by the upstream mt32emu.

**Comment by kcgen** [comment]
> > README.md
> > MT-32 emulator | Yes＊ (libmt32emu 2.4.2) | **N/A**
> 
> In DosBOX MT32 is supported via a patch provided by the upstream mt32emu.

This is true for almost all of Staging's major feature "checkboxes" that can similarly be added to DOSBox by way of community authored patches, which continue to get periodic re-syncs (and shared on vogons or the soueceforge patch list). 

Even features that are unique to staging can be diff'ed using git, and patches also generated against DOSBox proper - so ultimately DOSBox + *patches" can always equal and even exceed Staging in the list of features. 

That's all to say : I agree with your point, and will clarify the README.md wording that the comparison is merely against release binaries and unpatched source code.  (And once patches are added to the comparison, the sky is the limit 😅).

**Comment by IlyaIndigo** [comment]
Well.

**Comment by LowLevelMahn** [comment]
> QIf you're building from source, then I suggest using the current master branch (`git checkout master -f && git pull`), which gives you detailed model selection in the conf file.
>
> This is also described inside the config file it generates for you, if you delete your current one.

im using master - updated/rebuilded 5min ago :)
i've renamed my current conf but there is no new one generated - and starting dosbox without config disables my debugger keys like F5, PgUp/Down etc. - how to force clean config generation?

> Curious how it goes!

i try to understand the stunts sound bug - stated in the wiki link you've posted - thanks for the other links/infos very helpfull for testing

**Comment by kcgen** [comment]
> renamed my current conf but there is no new one generated 

Hmm; try running your executable with the `-editconf` argument.  If a conf file exists, it will open it in notepad and reveal its name/location. 

When you launch, inspect to top rows of your console output, which will also tell you the full path to the conf file being used. 

(If you're running in a game directory having a typical "dosbox.conf" file, then it will pick that up.. so launch in a directory without any local conf, just to help you find your 'primary' conf file).

**Comment by kcgen** [comment]
The `-writeconf` will write a new conf, and I believe it also accepts an optional new filename. `-writeconf dosbox.conf`. (but hopefully my prior comment helped you find/generate a new primary conf in your user account directory).

**Comment by LowLevelMahn** [comment]
works perfect - first time in years that i've got a brand new config - now my dosbox window is very small :) but thats ok

the all-in-one-folder + model-selections + indentification on console working perfect

thanks

**Comment by LowLevelMahn** [comment]
is see a "CONFIG: unknown option soundfont" with a fresh generated config, am i missing something or totaly unrelated to MT32 sound?

**Comment by Jkapp76** [comment]
I have never got the MT32 device music to play on any DOSBox emulator. I have the roms in a folder and named as you said and the conf file pointing to the directory, but the music all sounds like piano keys. sounds the same if I just delete the roms folder too.

**Comment by LowLevelMahn** [comment]
post your config 
the path to your .bin/.rom files
and a overview of the files in your rom-file folder

**Comment by Jkapp76** [comment]
I got it! I needed midi set to mt32. I missed the first setting above the mt32 area. Sounds perfect. 

⁣Get BlueMail for Android ​

On Feb 21, 2022, 12:34 AM, at 12:34 AM, LowLevelMahn ***@***.***> wrote:
>post your config 
>the path to your .bin/.rom files
>and a overview of the files in your rom-file folder
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/dosbox-staging/dosbox-staging/issues/257#issuecomment-1046491475
>You are receiving this because you commented.
>
>Message ID:
***@***.***>
